### PR TITLE
feat: add @eslint/v9-to-v10-ruletester codemod

### PR DESCRIPTION
**Pull request by khadatkarrohit**
## Summary

ESLint v10 added strict validation to `RuleTester` that throws on properties which were silently ignored in v9. Any test suite that passes `errors` or `output` in a `valid` case, or `type` in an `invalid` case, will crash after upgrading to v10.

This PR adds a new codemod `@eslint/v9-to-v10-ruletester` that automatically removes those forbidden properties so plugin authors can upgrade without touching their test files manually.

## What it transforms

**`valid` cases — remove `errors` and `output`**

```js
// Before
ruleTester.run('my-rule', rule, {
  valid: [
    { code: 'var x = 1', errors: [], output: null },
  ],
})

// After
ruleTester.run('my-rule', rule, {
  valid: [
    { code: 'var x = 1' },
  ],
})
```

**`invalid` cases — remove `type`**

```js
// Before
ruleTester.run('my-rule', rule, {
  invalid: [
    { code: 'eval(x)', type: 'CallExpression', errors: [{ messageId: 'noEval' }] },
  ],
})

// After
ruleTester.run('my-rule', rule, {
  invalid: [
    { code: 'eval(x)', errors: [{ messageId: 'noEval' }] },
  ],
})
```

## How to use

```bash
npx codemod @eslint/v9-to-v10-ruletester
```

## Testing

2 fixtures covering all transform cases — `errors`/`output` removal from valid cases and `type` removal from invalid cases.

```bash
pnpm --filter @eslint/v9-to-v10-ruletester test
pnpm --filter @eslint/v9-to-v10-ruletester check-types
```

All tests pass. No regressions in existing codemods (`pnpm run ci` green).

## Related

- Covers two breaking changes from the [ESLint v10 migration guide](https://eslint.org/docs/latest/use/migrate-to-10.0.0):
  - [Removal of `type` in invalid RuleTester cases](https://eslint.org/docs/latest/use/migrate-to-10.0.0#removal-of-type-property-in-errors-of-invalid-ruletester-cases)
  - [Prohibiting `errors`/`output` in valid RuleTester cases](https://eslint.org/docs/latest/use/migrate-to-10.0.0#prohibiting-errors-or-output-of-valid-ruletester-test-cases)
- Companion to #8 (`@eslint/v9-to-v10-custom-rules`) — both target ESLint plugin authors upgrading from v9 to v10

### CODE REVIEW

File: codemods/v10/ruletester/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# @eslint/v9-to-v10-ruletester
+
+Fix RuleTester test case structure for ESLint v10.
+
+## Overview
+
+> [!CAUTION]
+> Run this codemod only in directories containing ESLint rule test files. It removes properties from `RuleTester` test case objects — running it on unrelated code risks false positives if those properties appear inside `valid` or `invalid` arrays in other contexts.
```
Comment (alexbit-codemod) [inline codemods/v10/ruletester/README.md]: Ideally users should be able to safely run a codemod against the whole repo and codemod should be smart enough not to change irrelevant files. 

I think this is being replicated in these codemods because we had this issue in v8 to v9 codemod. 

Should we fix them all? @mohebifar @amirabbas-gh @khadatkarrohit

File: codemods/v10/ruletester/scripts/cleanup-valid-cases.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import { type RuleConfig, type SgNode, type SgRoot } from 'codemod:ast-grep'
+import type JS from 'codemod:ast-grep/langs/javascript'
+
+// Properties that v10 throws on when present in a valid test case
+const PROPS_TO_REMOVE = ['errors', 'output']
+
+function getSelector(): RuleConfig<JS> {
```
Comment (alexbit-codemod) [inline codemods/v10/ruletester/scripts/cleanup-valid-cases.ts]: Are these enough to make sure the code is only related to EsLint? Any other specifier we should use here? Like patterns that are inside ruleTester?

File: codemods/v10/ruletester/README.md (hunk 1 above)
Comment (khadatkarrohit) [reply codemods/v10/ruletester/README.md]: Agreed. The warning indicates the selector is too broad, not something users should manage. I’ll narrow it to only match inside .run() calls. Once that’s in place, the CAUTION block can be removed. I’ll push an update.

File: codemods/v10/ruletester/scripts/cleanup-valid-cases.ts (hunk 2 above)
Comment (khadatkarrohit) [reply codemods/v10/ruletester/scripts/cleanup-valid-cases.ts]: Good point. errors, output, and valid are generic enough to appear in non-ESLint code (form validators, compiler tests, etc.). I'll add an ancestor constraint scoping both scripts to only fire inside a .run() call, which is the reliable RuleTester-specific marker. Will push an update.

**Comment by khadatkarrohit** [comment]
Here is the [testing result](https://github.com/khadatkarrohit/eslint-v9-ruletester-test/blob/main/SCENARIOS-ruletester.md) for ruletester

File: codemods/v10/ruletester/scripts/cleanup-valid-cases.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  return {
+    rule: {
+      kind: 'pair',
+      has: {
+        kind: 'property_identifier',
+        regex: `^(${PROPS_TO_REMOVE.join('|')})$`,
+      },
```
[the remark this answers is not in the document]
Comment (khadatkarrohit) [reply codemods/v10/ruletester/scripts/cleanup-valid-cases.ts]: RuleTester test cases are always written with unquoted properties — { code: '...', errors: [], output: null }. No real ESLint plugin uses string-literal keys like { 'errors': [] } in test cases. Supporting quoted keys would also risk false positives on unrelated code (form validators, build configs) that happen to have an errors or output property. The property_identifier constraint is intentional and sufficient for this use case. No changes needed.

File: codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import type JS from 'codemod:ast-grep/langs/javascript'
+
+function getSelector(): RuleConfig<JS> {
+  return {
+    rule: {
+      kind: 'pair',
+      has: {
+        kind: 'property_identifier',
+        regex: '^type$',
+      },
```
[the remark this answers is not in the document]
Comment (khadatkarrohit) [reply codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts]: Valid observation, but type as a string-literal key in a RuleTester invalid case ({ 'type': 'CallExpression' }) doesn't exist in any real plugin — it's always unquoted. More importantly, type is an extremely common property name across JavaScript (Redux actions, GeoJSON, package.json). Matching quoted type keys would cause false positives on a lot of unrelated code. The property_identifier constraint is intentional to limit the match to the known RuleTester pattern only. No changes needed.

File: codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
+import { type RuleConfig, type SgNode, type SgRoot } from 'codemod:ast-grep'
+import type JS from 'codemod:ast-grep/langs/javascript'
+
+function getSelector(): RuleConfig<JS> {
+  return {
+    rule: {
+      kind: 'pair',
+      has: {
+        kind: 'property_identifier',
+        regex: '^type$',
```
Comment (DMartens) [inline codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts]: The `type` property is part of an error matcher inside `errors` (below "ERROR") not of an invalid test case (below "TESTCASE"):
```javascript
ruleTester.run("rule", rule, {
	invalid: [{ type: 'TESTCASE', errors: [{ type: 'ERROR' }]
});
```

Comment (khadatkarrohit) [reply codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts]: Fixed.

File: codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts [hunk 6]
```diff
@@ -9,6 +9,18 @@ function getSelector(): RuleConfig<JS> {
         kind: 'property_identifier',
         regex: '^type$',
       },
+      not: {
```
Comment (DMartens) [inline codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts]: You seem to have misunderstood me.
The `type` property was never allowed in a test case, only as an error matcher inside `errors`.
As such the codemod should only remove `type` properties inside `errors`.

Comment (khadatkarrohit) [reply codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts]: Thanks for the clarification, DMartens. I had misunderstood the original comment.

I’ve updated the selector to target type only within errors[] entries in invalid test cases and updated the fixtures to cover both removal and preservation cases.

File: codemods/v10/ruletester/workflow.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      - id: cleanup-valid-cases
+        name: 'Remove invalid properties from valid test cases'
+        js-ast-grep:
+          js_file: scripts/cleanup-valid-cases.ts
+          include:
+            - '**/*.ts'
+            - '**/*.js'
+            - '**/*.jsx'
+            - '**/*.cjs'
+            - '**/*.mjs'
```
Comment (lumirlumir) [inline codemods/v10/ruletester/workflow.yaml]:
```suggestion
            - '**/*.js'
            - '**/*.mjs'
            - '**/*.cjs'
            - '**/*.jsx' # Unlikely to be used for rule tester, but just in case
            - '**/*.ts'
            - '**/*.mts'
            - '**/*.cts'
            - '**/*.tsx' # Unlikely to be used for rule tester, but just in case
```

I think we need to modify the `include` pattern as above, since `RuleTester`s are not placed with JSX syntax and can usually be placed using the patterns above.

If adding the `jsx` extension is preferred just in case, I think adding the `tsx` extension would also make sense here.

File: codemods/v10/ruletester/workflow.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        depends_on:
+          - cleanup-valid-cases
+        js-ast-grep:
+          js_file: scripts/cleanup-invalid-cases.ts
+          include:
+            - '**/*.ts'
+            - '**/*.js'
+            - '**/*.jsx'
+            - '**/*.cjs'
+            - '**/*.mjs'
```
Comment (lumirlumir) [inline codemods/v10/ruletester/workflow.yaml]:
```suggestion
            - '**/*.js'
            - '**/*.mjs'
            - '**/*.cjs'
            - '**/*.jsx' # Unlikely to be used for rule tester, but just in case
            - '**/*.ts'
            - '**/*.mts'
            - '**/*.cts'
            - '**/*.tsx' # Unlikely to be used for rule tester, but just in case
```

Same here.

File: codemods/v10/ruletester/workflow.yaml (hunk 7 above)
Comment (lumirlumir) [inline codemods/v10/ruletester/workflow.yaml:16]:
```suggestion
            - '**/*.mjs'
          exclude:
            - '**/*.d.ts'
            - '**/*.d.cts'
            - '**/*.d.mts'
```

I also think TypeScript declaration files should be excluded, since they aren’t a place where `RuleTester` can be placed.

File: codemods/v10/ruletester/workflow.yaml (hunk 8 above)
Comment (lumirlumir) [inline codemods/v10/ruletester/workflow.yaml:35]:
```suggestion
            - '**/*.mjs'
          exclude:
            - '**/*.d.ts'
            - '**/*.d.cts'
            - '**/*.d.mts'
```

Same here.

File: codemods/v10/ruletester/README.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
+# @eslint/v9-to-v10-ruletester
+
+Fix RuleTester test case structure for ESLint v10.
```
Comment (lumirlumir) [inline codemods/v10/ruletester/README.md]:
```suggestion
Fix `RuleTester` test case structure for ESLint v10.
```

Non-blocking suggestion regarding inline code usage.

**Review by lumirlumir — CHANGES_REQUESTED** [review]

File: codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
+import { type RuleConfig, type SgNode, type SgRoot } from 'codemod:ast-grep'
+import type JS from 'codemod:ast-grep/langs/javascript'
+
+function getSelector(): RuleConfig<JS> {
```
Comment (lumirlumir) [inline codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts]: I think using a simple object form instead of the function form is fine here?

Since it doesn’t pass arguments, I think it would be cleaner and faster to use a simple object form with `as const satisfies RuleConfig` to indicate that it’s readonly.

Currently, the `getSelector()` function is called every time we call the `findAll` method, so the object is created every time, which would be a small but unnecessary overhead.

```ts
const selector = {
  rule: {
    kind: 'pair',
    has: {
      kind: 'property_identifier',
      regex: '^type$',
    },
    inside: {
      kind: 'object',
      inside: {
        kind: 'array',
        inside: {
          kind: 'pair',
          has: {
            kind: 'property_identifier',
            regex: '^errors$',
          },
          inside: {
            kind: 'object',
            inside: {
              kind: 'array',
              inside: {
                kind: 'pair',
                has: {
                  kind: 'property_identifier',
                  regex: '^invalid$',
                },
                inside: {
                  kind: 'object',
                  inside: {
                    kind: 'arguments',
                    inside: {
                      kind: 'call_expression',
                      has: {
                        kind: 'member_expression',
                        has: {
                          kind: 'property_identifier',
                          regex: '^run$',
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
} as const satisfies RuleConfig<JS>
```

File: codemods/v10/ruletester/scripts/cleanup-valid-cases.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
+import { type RuleConfig, type SgNode, type SgRoot } from 'codemod:ast-grep'
+import type JS from 'codemod:ast-grep/langs/javascript'
+
+// Properties that v10 throws on when present in a valid test case
+const PROPS_TO_REMOVE = ['errors', 'output']
+
+function getSelector(): RuleConfig<JS> {
```
Comment (lumirlumir) [inline codemods/v10/ruletester/scripts/cleanup-valid-cases.ts]: Same here. Using an object form instead of a function would be nice.

File: codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+export default async function transform(root: SgRoot<JS>): Promise<string | null> {
+  const rootNode = root.root()
+  const pairs = rootNode.findAll(getSelector())
+
+  if (pairs.length === 0) return null
+
+  const source = rootNode.text()
+  const deletions = pairs.map((pair) => removePairRange(pair, source)).sort((a, b) => b.start - a.start)
```
Comment (lumirlumir) [inline codemods/v10/ruletester/scripts/cleanup-invalid-cases.ts]: Manually sorting deletion ranges with `sort((a, b) => b.start - a.start)` and then applying them via `slice()` is working around the usual index-shift problem: once an earlier range is removed, later source indexes no longer point to the same text.

Since this transform already runs on JSSG/ast-grep, we can avoid managing that manually by returning ast-grep `Edit` objects and applying them with `SgNode#commitEdits(edits)`. That keeps all edits based on the original source positions and lets ast-grep apply them together.

For example (suggested direction):

```ts
import { type Edit, type RuleConfig, type SgNode, type SgRoot } from 'codemod:ast-grep'
import type JS from 'codemod:ast-grep/langs/javascript'

const selector = {
  rule: {
    kind: 'pair',
    has: {
      kind: 'property_identifier',
      regex: '^type$',
    },
    inside: {
      kind: 'object',
      inside: {
        kind: 'array',
        inside: {
          kind: 'pair',
          has: {
            kind: 'property_identifier',
            regex: '^errors$',
          },
          inside: {
            kind: 'object',
            inside: {
              kind: 'array',
              inside: {
                kind: 'pair',
                has: {
                  kind: 'property_identifier',
                  regex: '^invalid$',
                },
                inside: {
                  kind: 'object',
                  inside: {
                    kind: 'arguments',
                    inside: {
                      kind: 'call_expression',
                      has: {
                        kind: 'member_expression',
                        has: {
                          kind: 'property_identifier',
                          regex: '^run$',
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
} as const satisfies RuleConfig<JS>

function removePairEdit(pair: SgNode<JS>, source: string): Edit {
  const range = pair.range()
  const start = range.start.index
  const end = range.end.index

  // Prefer removing the preceding comma so the trailing element keeps its comma
  const before = source.slice(0, start)
  const precedingComma = before.match(/,\s*$/)

  if (precedingComma) {
    return {
      startPos: start - precedingComma[0].length,
      endPos: end,
      insertedText: '',
    }
  }

  // First property — remove trailing comma instead
  const after = source.slice(end)
  const trailingComma = after.match(/^\s*,/)

  if (trailingComma) {
    return {
      startPos: start,
      endPos: trailingComma ? end + trailingComma[0].length : end,
      insertedText: '',
    }
  }

  return {
    startPos: start,
    endPos: end,
    insertedText: '',
  }
}

export default async function transform(root: SgRoot<JS>): Promise<string | null> {
  const rootNode = root.root()
  const pairs = rootNode.findAll(selector)

  if (pairs.length === 0) return null

  const source = rootNode.text()
  const edits = pairs.map((pair) => removePairEdit(pair, source))

  return rootNode.commitEdits(edits)
}
```

Then the transform can apply all deletions at once:

```ts
const edits = pairs.map((pair) => removePairEdit(pair, source))

return rootNode.commitEdits(edits)
```

This removes the manual reverse sorting and `slice()` loop, and keeps the implementation closer to the ast-grep editing model.

File: codemods/v10/ruletester/scripts/cleanup-valid-cases.ts [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+export default async function transform(root: SgRoot<JS>): Promise<string | null> {
+  const rootNode = root.root()
+  const pairs = rootNode.findAll(getSelector())
+
+  if (pairs.length === 0) return null
+
+  const source = rootNode.text()
+  const deletions = pairs.map((pair) => removePairRange(pair, source)).sort((a, b) => b.start - a.start)
```
Comment (lumirlumir) [inline codemods/v10/ruletester/scripts/cleanup-valid-cases.ts]: Same here. Let’s use the `rootNode.commitEdits(edits)` pattern here too.

(It would require changing the internal implementation of the `removePairRange` function, as well as renaming it to `removePairEdit`.)

File: codemods/v10/ruletester/scripts/cleanup-valid-cases.ts [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+// Properties that v10 throws on when present in a valid test case
+const PROPS_TO_REMOVE = ['errors', 'output']
+
+function getSelector(): RuleConfig<JS> {
+  return {
+    rule: {
+      kind: 'pair',
+      has: {
+        kind: 'property_identifier',
+        regex: `^(${PROPS_TO_REMOVE.join('|')})$`,
```
Comment (lumirlumir) [inline codemods/v10/ruletester/scripts/cleanup-valid-cases.ts]:
```suggestion
        regex: `^(?:${PROPS_TO_REMOVE.join('|')})$`,
```

Non-blocking suggestion: I think we aren’t using the regex capture result here, so using a non-capturing group would be nice here.

**Review by lumirlumir — CHANGES_REQUESTED** [review]

File: codemods/v10/ruletester/README.md [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+Or run locally from source:
+
+```bash
+npx codemod workflow run -w codemods/v10/ruletester/workflow.yaml
+```
+
+## Resources
+
+- [ESLint v10 migration guide](https://eslint.org/docs/latest/use/migrate-to-10.0.0)
+- [Removal of `type` in invalid RuleTester cases](https://eslint.org/docs/latest/use/migrate-to-10.0.0#removal-of-type-property-in-errors-of-invalid-ruletester-cases)
```
Comment (lumirlumir) [inline codemods/v10/ruletester/README.md]:
```suggestion
- [Removal of `type` in invalid RuleTester cases](https://eslint.org/docs/latest/use/migrate-to-10.0.0#ruletester-type-removed)
```

It seems that the link is invalid.

File: codemods/v10/ruletester/README.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+- [Prohibiting `errors`/`output` in valid RuleTester cases](https://eslint.org/docs/latest/use/migrate-to-10.0.0#prohibiting-errors-or-output-of-valid-ruletester-test-cases)
```
Comment (lumirlumir) [inline codemods/v10/ruletester/README.md]:
```suggestion
- [Prohibiting `errors`/`output` in valid RuleTester cases](https://eslint.org/docs/latest/use/migrate-to-10.0.0#stricter-rule-tester)
```

File: codemods/v10/ruletester/scripts/cleanup-valid-cases.ts
Comment (lumirlumir) [inline codemods/v10/ruletester/scripts/cleanup-valid-cases.ts:1]: I think adding edge case test fixtures for `cleanup-valid-cases` as well would be helpful here.

For example, a test case in the style of `codemods/v10/ruletester/tests/cleanup-invalid-preserve-errors-type/input.js` will be a nice addition.

**Review by lumirlumir — COMMENTED** [review]
This would be my final review for this PR.

File: codemods/v10/ruletester/scripts/cleanup-valid-cases.ts
Comment (khadatkarrohit) [reply codemods/v10/ruletester/scripts/cleanup-valid-cases.ts:1]: added few prod level edge cases.

File: codemods/v10/ruletester/tests/cleanup-valid-errors-first-prop/expected.js [hunk 17]
```diff
@@ -0,0 +1,7 @@
+// errors/output as the first property triggers the trailing-comma removal branch
+ruleTester.run('my-rule', rule, {
+  valid: [
+    {  code: 'var x = 1' },
+    {  code: 'var y = 2', options: [{ strict: true }] },
```
Comment (lumirlumir) [inline codemods/v10/ruletester/tests/cleanup-valid-errors-first-prop/expected.js]:
```suggestion
    { code: 'var x = 1' },
    { code: 'var y = 2', options: [{ strict: true }] },
```

Is there a reason two spaces are left? I think the implementation logic needs to be changed to make sure there is only one space there.

It seems this also applies to both `cleanup-invalid-cases` and `cleanup-valid-cases`.

File: .lintstagedrc.mjs [hunk 18]
```diff
@@ -6,7 +6,16 @@ function excludeLockfiles(files) {
 }
 
 export default {
-  '*.{ts,tsx,js,jsx,mts,mjs}': ['oxfmt --write', 'oxlint --type-aware --type-check --fix'],
+  /** @param {string[]} files */
+  '*.{ts,tsx,js,jsx,mts,mjs}': (files) => {
+    const formattable = files.filter((f) => !f.includes('/tests/'))
```
Comment (lumirlumir) [inline .lintstagedrc.mjs]: What does this change do? I think the `.oxfmtrc.json` file already excludes the `tests` path, so theoretically excluding `/tests/` has no effect here.

https://github.com/eslint/codemods/blob/e44e66d712a86287e81687cf40490633ccbeb2d6/.oxfmtrc.json#L9

Also, the `/tests/` path handling won’t work on Windows, as it uses POSIX `/` as the path separator.

**Review by lumirlumir — CHANGES_REQUESTED** [review]

File: .lintstagedrc.mjs (hunk 18 above)
Comment (khadatkarrohit) [reply .lintstagedrc.mjs]: You're right on both points. The /tests/ filter is redundant since .oxfmtrc.json already excludes test files, and the current check is not Windows-safe.

The main reason for introducing the function form was different: oxlint --type-aware --type-check fails when only JS files are staged, as there are no TypeScript files available for type inference. A commit containing only .js and .md files consistently reproduces the issue.

Comment (lumirlumir) [reply .lintstagedrc.mjs]: Ah, I see the problem now.

If so, how about splitting the lint-staged script into three groups? For example:

```
'*.{js,mjs,cjs,jsx,ts,mts,cts,tsx}': 'oxfmt --write',
'*.{js,mjs,cjs,jsx}': 'oxlint --fix',
'*.{ts,mts,cts,tsx}': 'oxlint --type-aware --type-check --fix',
```

Comment (khadatkarrohit) [reply .lintstagedrc.mjs]: This is clean and solves both problems elegantly, split into three groups so JS and TS get different oxlint flags.

File: codemods/v10/ruletester/tests/cleanup-valid-errors-first-prop/expected.js (hunk 17 above)
Comment (khadatkarrohit) [reply codemods/v10/ruletester/tests/cleanup-valid-errors-first-prop/expected.js]: Done.

**Review by alexbit-codemod — APPROVED** [review]

**Review by lumirlumir — APPROVED** [review]
LGTM, thanks!
